src/ruby/puppetserver-lib/puppet/server/http_response.rb

```diff
@@ -0,0 +1,52 @@
+require 'puppet/http/response'
+
+class Puppet::Server::HttpResponse < Puppet::HTTP::Response
+  attr_reader :url
+
+  def initialize(java_response, url)
+    @java_response = java_response
+    @url = url
+  end
+
+  def code
+    @java_response.get_status
+  end
+
+  def reason
+    @java_response.get_reason_phrase
+  end
+
+  def body
+    @java_response.get_body
+  end
+
+  def read_body
+    yield body
+  end
+
+  # The HTTP standard counts all codes 200-299 as successes
+  def success?
+    code > 199 && code < 300
+  end
+
+  def [](name)
+    @java_response.get_headers[name]
+  end
+
+  # Yield each header name and value. Returns an enumerator if no block is given.
+  def each_header
+    if block_given?
+      @java_response.get_headers.each do |name, value|
+        yield(name, value)
+      end
+    else
+      @java_response.get_headers.each
+    end
+  end
+
+  # Drain the response body
+  def drain
```

Comment:
It appears to be used when handling redirects, see this [commit message](https://github.com/puppetlabs/puppet/commit/deafcdec7463c900a26de1b322660635007c1cf5).

Reply:
I guess I still just don't know what is meant by `drain`.. And I don't really understand what it does, other than return true??

Reply:
I'd recommend asking @joshcooper for details. But speculation: from what I can gather doing some cursory research is that sometimes, if you don't explicitly read the body from a response, it doesn't get pulled out of the connection state (this _might_ only be an issue if you asked for the body to be streamed, but I'm not sure). So if you then reuse the connection (e.g. during a redirect where you immediately visit another URL before actually returning), the body from the previous connection is still sitting there, in front of the one you actually want. So this function asks for the whole body, to clear it out and free up the space for the body of the next response.

Reply:
so calling `body` has the effect of clearing it out of somewhere? interesting.....

Reply:
I don't think it does, but the response isn't a shared object. IIUC, this is more for ensuring the underlying connection has drained all of this response's body before being re-used by another request. I think this is important when streaming but I believe the apache library handles this for us in non-streaming mode (ie, I think the connection is drained when creating the java response object).

Reply:
Ah the `drain` method is needed in cases where the underlying http client supports persistent HTTP connections and we need to make sure the response body is fully read from the socket. If we don't do that, and we put the connection back in the pool, then the next HTTP request using the same connection/socket will read the (partial) response body from the previous request. One way this can occur is during an HTTP redirect, as it's not super clear what the HTTP server is supposed to send in the body for all of the different redirect codes. https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections.

Although puppetserver's http client doesn't support persistent connections, you can implement the `drain` method by calling `body`, which will ensure the response body is fully read, and then thrown away.

By making `drain` part of the API, then we can easily add persistent connection handling to puppetserver's httpclient in the future without breaking consumers of the API, such as puppetdb terminus, report processor, etc.